doc: thread: extend documentation about Thread Border Router (#4576)
nrfconnect/sdk-nrf: doc/nrf/ug_thread_tools.rst

#. Enter the :file:`ot-nrf528xx` directory:

   .. code-block:: console

      cd ot-nrf528xx

#. Install the OpenThread dependencies:

   .. code-block:: console

      ./script/bootstrap

#. Build the RCP example for the hardware platform and the transport of your choice:

   .. tabs::

      .. tab:: nRF52840 Dongle (USB transport)

         .. code-block:: console

            rm -rf build
            script/build nrf52840 USB_trans -DOT_BOOTLOADER=USB -DOT_THREAD_VERSION=1.2

      .. tab:: nRF52840 Development Kit (UART transport)

         .. code-block:: console

            rm -rf build
            script/build nrf52840 UART_trans -DOT_THREAD_VERSION=1.2

   ..

   This creates an RCP image at :file:`build/bin/ot-rcp`.
#. Convert the RCP image to hexadecimal format:

   .. code-block:: console

      arm-none-eabi-objcopy -O ihex build/bin/ot-rcp build/bin/ot-rcp.hex

#. Depending on the hardware platform, complete the following steps:

   .. tabs::

      .. tab:: nRF52840 Dongle (USB transport)

         a. Install nRF Util:

            .. code-block:: console

               python3 -m pip install -U nrfutil

            .. note::

>                The nRF Util version distributed officially through PyPI is not supported on Raspberry Pi.

Reviewer: Wait - are we telling users first how to install nrfutil, and then that this installation doesn't work and they need to use a different way?
Or am I missing something?

Reviewer 2: AFAIK nrfutil doesn't work on Raspbian, because of the one dependency wrong version called pc_ble_driver. I think that  by following this instruction after calling `python3 -m pip install -U nrfutil` we will have all necessary dependencies installed, but one of them is wrong, so we need to install it manually and switch version later on. @LuDuda, correct?

Author: That is correct. It's quite unlikely that someone will build RCP image on RaspberryPi rather than on Linux-based PC, and so I decided just to put a note below.

Reviewer: Hmm, a bit skeptical about that ... I read this in the way that you either use a Raspberry Pi for everything or a Linux PC for everything. But maybe that is clearer when you actually do it.

Reviewer 3: One option to mitigate this: make it clearer in the note that it concerns Raspbian:
```suggestion
               If you are using Raspberry Pi, the nRF Util version distributed officially through PyPI is not supported on Raspbian OS.
To install a compatible version, execute the following commands:
```

Author: Agreed F2F change to:

```
If you are using Raspberry Pi, the nRF Util version distributed officially through PyPI is not supported on Raspbian OS.

To install a compatible version, execute the following commands:
```

Reviewer: Actually:
```
If you are using a Raspberry Pi, the nRF Util version distributed officially through PyPI is not supported.
To install a compatible version on Raspbian OS, execute the following commands:
```

               .. code-block:: console

                  sudo apt-get -y install libusb-1.0-0-dev sed
                  pip3 install click crcmod ecdsa intelhex libusb1 piccata protobuf pyserial pyyaml tqdm pc_ble_driver_py pyspinel
                  pip3 install -U --no-dependencies nrfutil==6.0.1
                  export PATH="$HOME/.local/bin:$PATH"

         #. Generate the RCP firmware package:

            .. code-block:: console

               nrfutil pkg generate --hw-version 52 --sd-req=0x00 \
                --application build/bin/ot-rcp.hex --application-version 1 build/bin/ot-rcp.zip

         #. Connect the nRF52840 Dongle to the USB port.
         #. Press the **RESET** button on the dongle to put it into the DFU mode.
            The LED on the dongle starts blinking red.
         #. Install the RCP firmware package onto the dongle by running the following command, with ``/dev/ttyACM0`` replaced with the device node name of your nRF52840 Dongle:

            .. code-block:: console

               nrfutil dfu usb-serial -pkg build/bin/ot-rcp.zip -p /dev/ttyACM0

      .. tab:: nRF52840 Development Kit (UART transport)

>          a. Program the image using the nrfjprog utility:

Reviewer: Can we assume that nrfjprog is installed?

Author: I changed the sentence to be aligned with other usage of nrfutil in our documentation:

```
         a. Program the image using the nrfjprog utility (which is part of the `nRF Command Line Tools`_):
```

   .. code-block:: console

      docker pull openthread/otbr

#. Connect the radio co-processor that you configured in :ref:`ug_thread_tools_tbr_rcp` to the Border Router device.
#. Start the OpenThread Border Router container using the following command (in the last line, replace ``/dev/ttyACM0`` with the device node name of the OpenThread radio co-processor):

   .. code-block:: console

      sudo docker run -it --rm --privileged --name otbr --network otbr -p 8080:80 \
      --sysctl "net.ipv6.conf.all.disable_ipv6=0 net.ipv4.conf.all.forwarding=1 net.ipv6.conf.all.forwarding=1" \
      --volume /dev/ttyACM0:/dev/radio openthread/otbr --radio-url spinel+hdlc+uart:///dev/radio

#. Form the Thread network using one of the following options:

   * Follow the instruction in the `OpenThread Border Router Codelab tutorial step 2`_.
   * Open the ``http://localhost:8080/`` address in a web browser and choose :guilabel:`Form` from the menu.

     .. note::
        If you are using a Raspberry Pi without a screen, but you have a different device in the same network, you can start a web browser on that device and use the address of the Raspberry Pi instead of ``localhost``.

>    Note down the selected On-Mesh Prefix value.

Reviewer 3: ```suggestion
#. Note down the selected On-Mesh Prefix value.
    For example, ``fd11:22::/64``.
```

>    You will need it in the next step to configure routing between the networks.

Reviewer 3: I think this sentence is redundant and can be removed. Next step clarifies this.

> 

Reviewer 3: Blank line not needed.

> #. To ensure that packets addressed to devices in the Thread network are routed through the OpenThread Border Router container in Docker,
>    run the following command with the On-Mesh Prefix that you configured in the previous step.
>    For example, for the ``fd11:22::/64`` prefix, enter the following command:

Reviewer 3: ```suggestion
#. Make sure that packets addressed to devices in the Thread network are routed through the OpenThread Border Router container in Docker by running the following command that uses the On-Mesh Prefix that you configured in the previous step (in this case, ``fd11:22::/64``):
```

Thread Border Router is a specific type of Border Router device that provides connectivity from the IEEE 802.15.4 network to adjacent networks on other physical layers (such as Wi-Fi or Ethernet).
Border Routers provide services for devices within the IEEE 802.15.4 network, including routing services for off-network operations.

> `OpenThread Border Router`_ (OTBR) released by Google is an open-source implementation of the Thread Border Router recommended to use for development purpose.

Reviewer 2: Do we even need this whole "`OpenThread Border Router`_ (OTBR) released by Google ..." sentence? Few lines below we already have very similar sentence: "For development purposes, you can use `OpenThread Border Router`_ ...", so maybe it would be better to remove this part from here and rephrase the part below, as follows:
"For development purposes, you can use the `OpenThread Border Router`_, an open-source Border Router implementation released by Google, that you can set up either on your PC using Docker or on Raspberry Pi."?

> `OpenThread Border Router`_ (OTBR) released by Google is an open-source implementation of the Thread Border Router recommended to use for development purpose.

Author: @kkasperczyk-no I removed it in the latest push.

Typically, a Border Router solution consists of the following parts:

* Application based on the :ref:`thread_architectures_designs_cp_ncp` design or its :ref:`thread_architectures_designs_cp_rcp` variant compatible with the IEEE 802.15.4 standard.
  This application can be implemented, for example, on an nRF52 device.
* Host-side application, usually implemented on a more powerful device with incorporated Linux-based operating system.

The |NCS| does not provide a complete Thread Border Router solution.
For development purposes, you can use `OpenThread Border Router`_, an open-source Border Router implementation that you can set up either on your PC using Docker or on Raspberry Pi.

OpenThread Border Router is compatible with Nordic Semiconductor devices.

> It minimally supports the following functions:

Author: I changed to: "It implements a number of features, including".

* Bidirectional IP connectivity between Thread and Wi-Fi or Ethernet networks (or both).

> * Bidirectional service discovery using mDNS (on Wi-Fi or Ethernet link, or both) and SRP (on Thread network).

Reviewer 2: How about "Bidirectional DNS-based service discovery"?

> * Thread-over-infrastructure that merges Thread partitions over IP-based links.

Author: I removed it as it is not implemented fully on the FW side yet.

* External Thread commissioning (for example, using a mobile phone) to authenticate and join a Thread device to a Thread network.

You can configure OpenThread Border Router either manually on Raspberry Pi or using a Docker container, as described in the following sections.

> If you go for the Docker container option, you must :ref:`configure radio co-processor <ug_thread_tools_tbr_rcp>` first.

Reviewer 2: I agree with @ru-fu. Wouldn't be better to a little bit reorganize this document and place `Configuring radio co-processor` on the first place as pre-requisity for following parts? If this section is kind of duplicated in the OpenThread Codelabs guide, you could place a note like this to make User understand what's going on: "In case of choosing Raspberry Pi option, you will find this section duplicated by the `OpenThread Border Router Codelab tutorial`_ _Build and flash RCP firmware_ section. You can either omit the current section for the Raspberry Pi and follow the Codelab tutorial or do it now and later omit _Build and flash RCP firmware_ section."

> If you go for the Docker container option, you must :ref:`configure radio co-processor <ug_thread_tools_tbr_rcp>` first.

Author: After F2F discussion, we would always mandate to go with "configure radio co-processor" section and omit the duplicated section in the Codelabs.

            .. code-block:: console

               JLinkExe -device NRF52840_XXAA -if SWD -speed 4000 -autoconnect 1 -SelectEmuBySN *SEGGER_ID*
               J-Link>MSDDisable
               Probe configured successfully.
               J-Link>exit

            Replace *SEGGER_ID* with the SEGGER ID of your nRF52840 Development Kit.
            This setting remains valid even if you flash another firmware onto the device.
         #. Power-cycle the device to apply the changes.

Configuring using Docker
========================

For the development purposes, you can run the OpenThread Border Router on any Linux-based system using a Docker container that already has the Border Router installed.
This solution is better when you are only interested in direct communication between your Border Router and the Thread network.
For example, you can use the Docker container when you want to establish IP communication between an application running on Linux (such as Python Controller for Matter) and an application running on a Thread node.

If you go for the Docker container option, make sure to :ref:`configure radio co-processor <ug_thread_tools_tbr_rcp>` first.

> To install and configure OpenThread Border Router on any Linux-based operating system using the Docker container, complete the following steps:

Reviewer: It looks like these steps are for Ubuntu/Debian, not "any Linux-based operating system"?

Author: That is right, in general, the OTBR can be installed on any Linux-based distribution, but we show how to achieve it on Ubuntu. I changed the wording.

#. Install the Docker daemon:

   .. code-block:: console

      sudo apt update && sudo apt install docker.io

#. Start the Docker daemon:

   .. code-block:: console

      sudo systemctl start docker

#. Create an IPv6 network for the OpenThread Border Router container in Docker:

   .. code-block:: console

      sudo docker network create --ipv6 --subnet fd11:db8:1::/64 -o com.docker.network.bridge.name=otbr0 otbr

#. Download the latest version of the OpenThread Border Router Docker image by running the following command:

   .. code-block:: console

      docker pull openthread/otbr

> #. Connect the radio co-processor that you configured in :ref:`ug_thread_tools_tbr_rcp` to the Border Router device.

Reviewer: Thus your Linux device?

Author: Correct. Do you want me to add this additional clarification?

#. Start the OpenThread Border Router container using the following command (in the last line, replace */dev/ttyACM0* with the device node name of the OpenThread radio co-processor):

   .. parsed-literal::
      :class: highlight

      sudo docker run -it --rm --privileged --name otbr --network otbr -p 8080:80 \
      --sysctl "net.ipv6.conf.all.disable_ipv6=0 net.ipv4.conf.all.forwarding=1 net.ipv6.conf.all.forwarding=1" \
      --volume */dev/ttyACM0*:/dev/radio openthread/otbr --radio-url spinel+hdlc+uart:///dev/radio

#. Form the Thread network using one of the following options:

   * Follow the instruction in the `OpenThread Border Router Codelab tutorial step 2`_.
   * Open the ``http://*localhost*:8080/`` address in a web browser and choose :guilabel:`Form` from the menu.

> 

Reviewer 2: If we need to use On-Mesh Prefix set in this step later on, maybe it would be good idea to mention it here, to not force User to go back again. How about adding the following sentence: "Please note the selected On-Mesh Prefix value, as in further step it will be needed in order to configure routing between the networks."

     .. note::
        If you are using a Raspberry Pi without a screen, but you have a different device in the same network, you can start a web browser on that device and use the address of the Raspberry Pi instead of *localhost*.

#. To ensure that packets addressed to devices in the Thread network are routed through the OpenThread Border Router container in Docker,

>    run the following command with the On-Mesh Prefix that you set earlier for example, for ``fd11:22::/64`` prefix:

Reviewer: Might be good to specify that "that you set earlier" does not refer to the current set of steps, but that you set it when you configured the Thread network (correct?).

Author: It's actually the previous step where the user needs to "Form" the network.

   .. code-block:: console

      sudo ip -6 route add fd11:22::/64 dev otbr0 via fd11:db8:1::2

#. Check the status of the OpenThread Border Router by executing the following command:

   .. code-block:: console

      sudo docker exec -it otbr sh -c "sudo service otbr-agent status"

>       * otbr-agent is not running

Reviewer: Shouldn't it be running?

Author: I removed the output of the commands, to align with the rest of the documentation.
